File: android/app/src/main/java/com/nearme/app/ui/components/InputComponents.kt
package com.nearme.app.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.text.input.VisualTransformation
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import com.nearme.app.ui.theme.*

@Composable
fun TextInputField(
    label: String,
    value: String,
    onValueChange: (String) -> Unit,
    modifier: Modifier = Modifier,
    placeholder: String = "",
    isPassword: Boolean = false,
    keyboardType: KeyboardType = KeyboardType.Text,
    errorMessage: String? = null,
    isRequired: Boolean = false,
    isEnabled: Boolean = true
) {
    Column(
        modifier = modifier,
        verticalArrangement = Arrangement.spacedBy(DesignSystem.Spacing.xs)
    ) {
        Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(DesignSystem.Spacing.xs)
        ) {
            Text(
                text = label,
                style = MaterialTheme.typography.bodyMedium,
                color = TextPrimary,
                fontWeight = FontWeight.Medium
            )
            
            if (isRequired) {
                Text(
                    text = "*",
                    style = MaterialTheme.typography.bodyMedium,
                    color = Error
                )
            }
        }
        
        OutlinedTextField(
            value = value,
            onValueChange = onValueChange,
            placeholder = {
                Text(
                    text = placeholder,
                    color = TextSecondary
                )
            },
            enabled = isEnabled,
            singleLine = true,
            keyboardOptions = KeyboardOptions(keyboardType = keyboardType),
            visualTransformation = if (isPassword) PasswordVisualTransformation() else VisualTransformation.None,
            trailingIcon = if (isPassword) {
                IconButton(onClick = { /* Toggle password visibility */ }) {
                    Icon(
                        imageVector = Icons.Default.Visibility,
                        contentDescription = "Toggle password visibility",
                        tint = TextSecondary
                    )
                }
            } else null,
            colors = OutlinedTextFieldDefaults.colors(
                focusedBorderColor = Primary,
                unfocusedBorderColor = Border,
                focusedTextColor = TextPrimary,
                unfocusedTextColor = TextPrimary,
                disabledTextColor = TextTertiary,
                disabledBorderColor = Border.copy(alpha = 0.5f)
            ),
            shape = RoundedCornerShape(DesignSystem.CornerRadius.md),
            modifier = Modifier.fillMaxWidth()
        )
        
        if (errorMessage != null) {
            Text(
                text = errorMessage,
                style = MaterialTheme.typography.labelMedium,
                color = Error
            )
        }
    }
}

@Composable
fun SearchField(
    value: String,
    onValueChange: (String) -> Unit,
    modifier: Modifier = Modifier,
    placeholder: String = "Search...",
    onSearchClick: (() -> Unit)? = null
) {
    var isEditing by remember { mutableStateOf(false) }
    
    OutlinedTextField(
        value = value,
        onValueChange = { newValue ->
            onValueChange(newValue)
            isEditing = newValue.isNotEmpty()
        },
        placeholder = {
            Text(
                text = placeholder,
                color = TextSecondary
            )
        },
        leadingIcon = {
            Icon(
                imageVector = Icons.Default.Search,
                contentDescription = null,
                tint = TextSecondary
            )
        },
        trailingIcon = {
            if (isEditing && value.isNotEmpty()) {
                IconButton(onClick = {
                    onValueChange("")
                    isEditing = false
                }) {
                    Icon(
                        imageVector = Icons.Default.Close,
                        contentDescription = "Clear search",
                        tint = TextSecondary
                    )
                }
            }
        },
        singleLine = true,
        colors = OutlinedTextFieldDefaults.colors(
            focusedBorderColor = Primary,
            unfocusedBorderColor = Border,
            focusedTextColor = TextPrimary,
            unfocusedTextColor = TextPrimary
        ),
        shape = RoundedCornerShape(DesignSystem.CornerRadius.md),
        modifier = modifier.fillMaxWidth()
    )
}

@Composable
fun <T> PickerField(
    label: String,
    selectedValue: T,
    onValueChange: (T) -> Unit,
    options: List<T>,
    optionLabel: (T) -> String,
    modifier: Modifier = Modifier,
    isRequired: Boolean = false,
    errorMessage: String? = null,
    isEnabled: Boolean = true
) {
    var isExpanded by remember { mutableStateOf(false) }
    
    Column(
        modifier = modifier,
        verticalArrangement = Arrangement.spacedBy(DesignSystem.Spacing.xs)
    ) {
        Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(DesignSystem.Spacing.xs)
        ) {
            Text(
                text = label,
                style = MaterialTheme.typography.bodyMedium,
                color = TextPrimary,
                fontWeight = FontWeight.Medium
            )
            
            if (isRequired) {
                Text(
                    text = "*",
                    style = MaterialTheme.typography.bodyMedium,
                    color = Error
                )
            }
        }
        
        ExposedDropdownMenuBox(
            expanded = isExpanded,
            onExpandedChange = { isExpanded = !isExpanded },
            modifier = Modifier.fillMaxWidth()
        ) {
            OutlinedTextField(
                value = optionLabel(selectedValue),
                onValueChange = {},
                readOnly = true,
                enabled = isEnabled,
                trailingIcon = {
                    ExposedDropdownMenuDefaults.TrailingIcon(expanded = isExpanded)
                },
                colors = OutlinedTextFieldDefaults.colors(
                    focusedBorderColor = Primary,
                    unfocusedBorderColor = Border,
                    focusedTextColor = TextPrimary,
                    unfocusedTextColor = TextPrimary,
                    disabledTextColor = TextTertiary,
                    disabledBorderColor = Border.copy(alpha = 0.5f)
                ),
                shape = RoundedCornerShape(DesignSystem.CornerRadius.md),
                modifier = Modifier
                    .fillMaxWidth()
                    .menuAnchor()
            )
            
            ExposedDropdownMenu(
                expanded = isExpanded,
                onDismissRequest = { isExpanded = false }
            ) {
                options.forEach { option ->
                    DropdownMenuItem(
                        text = {
                            Text(
                                text = optionLabel(option),
                                color = if (option == selectedValue) Primary else TextPrimary
                            )
                        },
                        onClick = {
                            onValueChange(option)
                            isExpanded = false
                        },
                        leadingIcon = if (option == selectedValue) {
                            {
                                Icon(
                                    imageVector = Icons.Default.Check,
                                    contentDescription = null,
                                    tint = Primary
                                )
                            }
                        } else null
                    )
                }
            }
        }
        
        if (errorMessage != null) {
            Text(
                text = errorMessage,
                style = MaterialTheme.typography.labelMedium,
                color = Error
            )
        }
    }
}

@Composable
fun ToggleField(
    title: String,
    checked: Boolean,
    onCheckedChange: (Boolean) -> Unit,
    modifier: Modifier = Modifier,
    description: String? = null
) {
    Row(
        modifier = modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.SpaceBetween,
        verticalAlignment = Alignment.CenterVertically
    ) {
        Column(
            modifier = Modifier.weight(1f),
            verticalArrangement = Arrangement.spacedBy(DesignSystem.Spacing.xs)
        ) {
            Text(
                text = title,
                style = MaterialTheme.typography.bodyMedium,
                color = TextPrimary,
                fontWeight = FontWeight.Medium
            )
            
            if (description != null) {
                Text(
                    text = description,
                    style = MaterialTheme.typography.labelMedium,
                    color = TextSecondary
                )
            }
        }
        
        Switch(
            checked = checked,
            onCheckedChange = onCheckedChange,
            colors = SwitchDefaults.colors(
                checkedThumbColor = TextInverse,
                checkedTrackColor = Primary,
                uncheckedThumbColor = TextInverse,
                uncheckedTrackColor = Border
            )
        )
    }
}

@Composable
fun SliderField(
    title: String,
    value: Float,
    onValueChange: (Float) -> Unit,
    valueRange: ClosedFloatingPointRange<Float>,
    modifier: Modifier = Modifier,
    unit: String = "",
    description: String? = null
) {
    Column(
        modifier = modifier,
        verticalArrangement = Arrangement.spacedBy(DesignSystem.Spacing.sm)
    ) {
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically
        ) {
            Text(
                text = title,
                style = MaterialTheme.typography.bodyMedium,
                color = TextPrimary,
                fontWeight = FontWeight.Medium
            )
            
            Text(
                text = "${value.toInt()}${unit}",
                style = MaterialTheme.typography.bodyMedium,
                color = Primary,
                fontWeight = FontWeight.Medium
            )
        }
        
        if (description != null) {
            Text(
                text = description,
                style = MaterialTheme.typography.labelMedium,
                color = TextSecondary
            )
        }
        
        Slider(
            value = value,
            onValueChange = onValueChange,
            valueRange = valueRange,
            colors = SliderDefaults.colors(
                thumbColor = Primary,
                activeTrackColor = Primary,
                inactiveTrackColor = Border
            )
        )
    }
}

@Preview(showBackground = true)
@Composable
fun InputComponentsPreview() {
    NearMeTheme {
        Column(
            modifier = Modifier.padding(DesignSystem.Spacing.lg),
            verticalArrangement = Arrangement.spacedBy(DesignSystem.Spacing.lg)
        ) {
            TextInputField(
                label = "Task Title",
                value = "",
                onValueChange = {},
                placeholder = "Enter task title",
                isRequired = true
            )
            
            SearchField(
                value = "",
                onValueChange = {},
                placeholder = "Search places..."
            )
            
            PickerField(
                label = "Category",
                selectedValue = "Option 1",
                onValueChange = {},
                options = listOf("Option 1", "Option 2", "Option 3"),
                optionLabel = { it }
            )
            
            ToggleField(
                title = "Enable Notifications",
                checked = false,
                onCheckedChange = {},
                description = "Receive notifications when near your tasks"
            )
            
            SliderField(
                title = "Radius",
                value = 5f,
                onValueChange = {},
                valueRange = 0.1f..10f,
                unit = " mi",
                description = "Distance from location to trigger notification"
            )
        }
    }
}

File: NearMe/Components/InputComponents.swift
import SwiftUI

// MARK: - Input Components
struct TextInputField: View {
    let title: String
    @Binding var text: String
    var placeholder: String = ""
    var isSecure: Bool = false
    var keyboardType: UIKeyboardType = .default
    var errorMessage: String?
    var isRequired: Bool = false
    
    @State private var isFocused: Bool = false
    
    var body: some View {
        VStack(alignment: .leading, spacing: DesignSystem.Spacing.xs) {
            HStack {
                Text(title)
                    .font(DesignSystem.Typography.body)
                    .fontWeight(.medium)
                    .foregroundColor(DesignSystem.Colors.textPrimary)
                
                if isRequired {
                    Text("*")
                        .foregroundColor(DesignSystem.Colors.error)
                }
            }
            
            Group {
                if isSecure {
                    SecureField(placeholder, text: $text)
                } else {
                    TextField(placeholder, text: $text)
                        .keyboardType(keyboardType)
                }
            }
            .font(DesignSystem.Typography.body)
            .padding(DesignSystem.Spacing.md)
            .background(
                RoundedRectangle(cornerRadius: DesignSystem.CornerRadius.md)
                    .fill(DesignSystem.Colors.surface)
                    .overlay(
                        RoundedRectangle(cornerRadius: DesignSystem.CornerRadius.md)
                            .stroke(
                                isFocused ? DesignSystem.Colors.primary : DesignSystem.Colors.border,
                                lineWidth: isFocused ? 2 : 1
                            )
                    )
            )
            .foregroundColor(DesignSystem.Colors.textPrimary)
            .onTapGesture {
                isFocused = true
            }
            .onChange(of: text) { _ in
                isFocused = false
            }
            
            if let errorMessage = errorMessage {
                Text(errorMessage)
                    .font(DesignSystem.Typography.caption1)
                    .foregroundColor(DesignSystem.Colors.error)
            }
        }
    }
}

struct SearchField: View {
    @Binding var text: String
    var placeholder: String = "Search..."
    var onSearchButtonClicked: (() -> Void)?
    
    @State private var isEditing: Bool = false
    
    var body: some View {
        HStack(spacing: DesignSystem.Spacing.sm) {
            Image(systemName: "magnifyingglass")
                .foregroundColor(DesignSystem.Colors.textSecondary)
                .font(.system(size: 16, weight: .medium))
            
            TextField(placeholder, text: $text)
                .font(DesignSystem.Typography.body)
                .foregroundColor(DesignSystem.Colors.textPrimary)
                .onTapGesture {
                    isEditing = true
                }
                .onChange(of: text) { _ in
                    if text.isEmpty {
                        isEditing = false
                    }
                }
            
            if isEditing && !text.isEmpty {
                Button(action: {
                    text = ""
                    isEditing = false
                }) {
                    Image(systemName: "xmark.circle.fill")
                        .foregroundColor(DesignSystem.Colors.textSecondary)
                        .font(.system(size: 16))
                }
            }
        }
        .padding(DesignSystem.Spacing.md)
        .background(
            RoundedRectangle(cornerRadius: DesignSystem.CornerRadius.md)
                .fill(DesignSystem.Colors.surface)
                .overlay(
                    RoundedRectangle(cornerRadius: DesignSystem.CornerRadius.md)
                        .stroke(DesignSystem.Colors.border, lineWidth: 1)
                )
        )
    }
}

struct PickerField<T: Hashable>: View {
    let title: String
    @Binding var selection: T
    let options: [T]
    let optionTitle: (T) -> String
    var isRequired: Bool = false
    var errorMessage: String?
    
    @State private var isExpanded: Bool = false
    
    var body: some View {
        VStack(alignment: .leading, spacing: DesignSystem.Spacing.xs) {
            HStack {
                Text(title)
                    .font(DesignSystem.Typography.body)
                    .fontWeight(.medium)
                    .foregroundColor(DesignSystem.Colors.textPrimary)
                
                if isRequired {
                    Text("*")
                        .foregroundColor(DesignSystem.Colors.error)
                }
            }
            
            Button(action: {
                withAnimation(DesignSystem.Animation.fast) {
                    isExpanded.toggle()
                }
            }) {
                HStack {
                    Text(optionTitle(selection))
                        .font(DesignSystem.Typography.body)
                        .foregroundColor(DesignSystem.Colors.textPrimary)
                        .frame(maxWidth: .infinity, alignment: .leading)
                    
                    Image(systemName: "chevron.down")
                        .foregroundColor(DesignSystem.Colors.textSecondary)
                        .font(.system(size: 14, weight: .medium))
                        .rotationEffect(.degrees(isExpanded ? 180 : 0))
                        .animation(DesignSystem.Animation.fast, value: isExpanded)
                }
                .padding(DesignSystem.Spacing.md)
                .background(
                    RoundedRectangle(cornerRadius: DesignSystem.CornerRadius.md)
                        .fill(DesignSystem.Colors.surface)
                        .overlay(
                            RoundedRectangle(cornerRadius: DesignSystem.CornerRadius.md)
                                .stroke(DesignSystem.Colors.border, lineWidth: 1)
                        )
                )
            }
            
            if isExpanded {
                VStack(spacing: 0) {
                    ForEach(options, id: \.self) { option in
                        Button(action: {
                            selection = option
                            withAnimation(DesignSystem.Animation.fast) {
                                isExpanded = false
                            }
                        }) {
                            HStack {
                                Text(optionTitle(option))
                                    .font(DesignSystem.Typography.body)
                                    .foregroundColor(DesignSystem.Colors.textPrimary)
                                    .frame(maxWidth: .infinity, alignment: .leading)
                                
                                if selection == option {
                                    Image(systemName: "checkmark")
                                        .foregroundColor(DesignSystem.Colors.primary)
                                        .font(.system(size: 14, weight: .medium))
                                }
                            }
                            .padding(DesignSystem.Spacing.md)
                            .background(
                                Rectangle()
                                    .fill(selection == option ? DesignSystem.Colors.primary.opacity(0.1) : DesignSystem.Colors.background)
                            )
                        }
                    }
                }
                .background(
                    RoundedRectangle(cornerRadius: DesignSystem.CornerRadius.md)
                        .fill(DesignSystem.Colors.background)
                        .overlay(
                            RoundedRectangle(cornerRadius: DesignSystem.CornerRadius.md)
                                .stroke(DesignSystem.Colors.border, lineWidth: 1)
                        )
                )
                .designSystemShadow(DesignSystem.Shadow.small)
            }
            
            if let errorMessage = errorMessage {
                Text(errorMessage)
                    .font(DesignSystem.Typography.caption1)
                    .foregroundColor(DesignSystem.Colors.error)
            }
        }
    }
}

struct ToggleField: View {
    let title: String
    @Binding var isOn: Bool
    var description: String?
    
    var body: some View {
        HStack {
            VStack(alignment: .leading, spacing: DesignSystem.Spacing.xs) {
                Text(title)
                    .font(DesignSystem.Typography.body)
                    .fontWeight(.medium)
                    .foregroundColor(DesignSystem.Colors.textPrimary)
                
                if let description = description {
                    Text(description)
                        .font(DesignSystem.Typography.caption1)
                        .foregroundColor(DesignSystem.Colors.textSecondary)
                }
            }
            
            Spacer()
            
            Toggle("", isOn: $isOn)
                .toggleStyle(SwitchToggleStyle(tint: DesignSystem.Colors.primary))
        }
        .padding(DesignSystem.Spacing.md)
        .background(
            RoundedRectangle(cornerRadius: DesignSystem.CornerRadius.md)
                .fill(DesignSystem.Colors.surface)
        )
    }
}

struct SliderField: View {
    let title: String
    @Binding var value: Double
    let range: ClosedRange<Double>
    let step: Double
    var unit: String = ""
    var description: String?
    
    var body: some View {
        VStack(alignment: .leading, spacing: DesignSystem.Spacing.sm) {
            HStack {
                Text(title)
                    .font(DesignSystem.Typography.body)
                    .fontWeight(.medium)
                    .foregroundColor(DesignSystem.Colors.textPrimary)
                
                Spacer()
                
                Text("\(Int(value))\(unit)")
                    .font(DesignSystem.Typography.body)
                    .fontWeight(.medium)
                    .foregroundColor(DesignSystem.Colors.primary)
            }
            
            if let description = description {
                Text(description)
                    .font(DesignSystem.Typography.caption1)
                    .foregroundColor(DesignSystem.Colors.textSecondary)
            }
            
            Slider(value: $value, in: range, step: step)
                .accentColor(DesignSystem.Colors.primary)
        }
        .padding(DesignSystem.Spacing.md)
        .background(
            RoundedRectangle(cornerRadius: DesignSystem.CornerRadius.md)
                .fill(DesignSystem.Colors.surface)
        )
    }
}

// MARK: - Input Previews
struct InputComponents_Previews: PreviewProvider {
    @State static var text = ""
    @State static var searchText = ""
    @State static var selectedOption = "Option 1"
    @State static var isToggleOn = false
    @State static var sliderValue = 5.0
    
    static var previews: some View {
        ScrollView {
            VStack(spacing: DesignSystem.Spacing.lg) {
                TextInputField(
                    title: "Task Title",
                    text: $text,
                    placeholder: "Enter task title",
                    isRequired: true
                )
                
                SearchField(text: $searchText, placeholder: "Search places...")
                
                PickerField(
                    title: "Category",
                    selection: $selectedOption,
                    options: ["Option 1", "Option 2", "Option 3"],
                    optionTitle: { $0 }
                )
                
                ToggleField(
                    title: "Enable Notifications",
                    isOn: $isToggleOn,
                    description: "Receive notifications when near your tasks"
                )
                
                SliderField(
                    title: "Radius",
                    value: $sliderValue,
                    range: 0.1...10.0,
                    step: 0.1,
                    unit: " mi",
                    description: "Distance from location to trigger notification"
                )
            }
            .padding()
        }
        .background(DesignSystem.Colors.background)
    }
}
